### ISSUE TITLE
[Enhancement] Applescript API

### ISSUE THREAD

**Issue by Coder-256** [electron/electron#4418]
Is it possible to have an Applescript API for Mac apps made with Electron?

**Comment by dkfiresky** [comment]
You can watch my repo https://github.com/over-powdered/soap

The module allows you to execute AppleScript from Node.JS

**Comment by bengotow** [comment]
I think this is probably best done through a native node module like the one above rather than being built into the core Electron framework.

**Comment by kevinsawicki** [comment]
> I think this is probably best done through a native node module like the one above rather than being built into the core Electron framework.

Agreed, going to close this out, thanks for sharing that link @dkfiresky

**Comment by Coder-256** [comment]
I was talking about something more like this: https://developer.apple.com/library/mac/documentation/AppleScript/Conceptual/AppleScriptX/Concepts/scriptable_apps.html

**Comment by nimish** [comment]
This is for support for exposing functionality inside the electron app to applescript, not calling applescript via the node app, so that things like beardedspice can work

**Comment by Coder-256** [comment]
What I meant was I wanted BeardedSpice to be able to execute Javascript in Electron apps. I was using AppleScript as a potential medium for doing this. On a Mac, open Script Editor and open `File > Open Dictionary...` and select basically any web browser. Choose its own suite on the left and you will se an `execute` or `executeJavascript` or similar function which **lets you run Javascript inside the browser from another application using Applescript**.

**Comment by dkfiresky** [comment]
@Coder-256 probably you need to implement this feature on your own as it's very complex.

**Comment by paulcbetts** [comment]
It's probably not reasonable to implement AppleScript support in Electron apps, because the format for Script Dictionaries is essentially a dictionary of "Script command => Objective C selector" mappings. 

_Theoretically_, you could map all of these to a Proxy object then marshal the call to a JS event (probably on `app`), but my Objective C-fu is lacking in this department. Maybe @joshaber has an idea?

**Comment by joshaber** [comment]
Oof, yeah, you could dynamically implement selectors using the Objective-C runtime to call through to JS.

**Comment by Coder-256** [comment]
@joshaber Seems interesting and complicated. How should I begin approaching this?

**Comment by joshaber** [comment]
https://developer.apple.com/library/ios/documentation/Cocoa/Conceptual/ObjCRuntimeGuide/Articles/ocrtDynamicResolution.html is probably a good starting point.

**Comment by aricwilmunder** [comment]
I'm trying to click menu items in an Electron App using AppleScript but I'm not seeing any elements when I open the apps dictionary.  Anyone with success in this area?

**Comment by capripot** [comment]
@kevinsawicki Out of curiosity, how come "touch bar" support (which is Mac only) is implemented directly within the electron project and AppleScript support would be implemented through a native node module?

**Comment by capripot** [comment]
Out of curiosity, how come [touchbar support](https://github.com/electron/electron/pull/8095) is built-in but AppleScript api should be a node module?

**Comment by dkfiresky** [comment]
@capripot touchbar support is application related thing, and AppleScript is not.
AppleScript could be run separately without any need of application kind of wrapper.
Touchbar, on the other hand, is part of the application framework, like, for example, the toolbar.

Image if every API provided by the macOS would be implemented in the electron's core. The core then will be like the whole operating system kind of thing.

**Comment by jhoughjr** [comment]
Just want to point out, AppleScript maps script commands to AppleEvents, not Objc selectors directly.  I gather the selectors are joined by the sdef file combined with KVC in NSObject.
In theory one should be able to consume and generate apple events without objc.

**Comment by gdsmith** [comment]
> Image if every API provided by the macOS would be implemented in the electron's core. The core then will be like the whole operating system kind of thing.

Maybe I'm not following the arguments correctly but I think there's a bit of a misunderstanding about what AppleScript support means (and how it could be used) that for me, puts it directly in line with TouchBar support. AppleScript support should be about opening up your application to an external scripting language.

Imagine the following AppleScript:

```AppleScript
tell application "MyElectronApp"
    open file "file:///path/to/a/file"
end tell
```

or

```AppleScript
property currentTrackName : "Unknown Track"

tell application "MyElectronApp"
    set currentTrackName to name of the current track
end tell
```

This is not about allowing your app to arbitrarily call out to AppleScript, it's rather _exposing specific functionality of the application_ to AppleScript.

Imagine the TouchBar game example, rewritten for AppleScript with a simple find/replace:

```
TouchBarLabel  => AppleScriptObject
TouchBarButton => AppleScriptCommand
TouchBar       => AppleScriptDictionary
```

Imagine if AppleScript support was available in Electron, aside from removing some layout specific code I'm not sure how much more you would have to do over the above renaming in order to have a functioning AppleScript version of that, enabling something like:

```AppleScript
tell application "MyElectronApp"
    spin reels
end tell
```

Please don't think I'm arguing that it would be easy, just that it falls into the realm of enabling user interaction in your application in exactly the same way as TouchBar support does.

**Comment by nornagon** [comment]
I agree that this issue was misinterpreted as "I want to be able to send AppleScript events to other applications", but the real intent was "I want to be able to set up my app to receive AppleScript events".

Looking through some Cocoa docs such as https://developer.apple.com/documentation/foundation/nsscriptsuiteregistry?language=objc, this seems not obviously impossible to do programmatically, so I'm reopening this issue as I think this is at least worth exploring.

**Comment by ThomasSzabo** [comment]
Has anyone had any success with receiving AppleScript events in an electron app?

**Comment by ThomasSzabo** [comment]
> Oof, yeah, you could dynamically implement selectors using the Objective-C runtime to call through to JS.

Is electron providing an JS APIs to dynamically implement Obj-C code at runtime? @joshaber how would you go about this?

**Comment by nornagon** [comment]
@ThomasSzabo we don't provide such a thing, but check out [objc](https://www.npmjs.com/package/objc).

**Comment by ThomasSzabo** [comment]
@nornagon this is nice. I also tried a native module and liked it a lot. But I am not getting how I can send an incoming AppleEvent (as defined in the sdef file) to the native node module binary that is hidden deep inside the app folder.

Is the conclusion of this thread that I have to build my own custom electron with AppleScript in order to receive AppleScript events?
Or are there any other ways? Has anyone accomplished anything?

**Comment by gdsmith** [comment]
bump

**Comment by codebykevin** [comment]
Implementing a full AppleScript API with the ability to specify objects and properties on an app-specific basis may be too complex. 

However, it appears that, whether intended or not, the core Apple Events and AppleScript commands -- launching the app, opening files, display dialogs, quitting the app -- are already present. This AppleScript runs without error:

`tell application "/Applications/Visual Studio Code.app"
	launch
	open POSIX file "/Users/kevin/Desktop/anigif.tcl"
	display dialog "quitting now..."
	delay 3
	quit
end tell`

This script launches VS Code, opens the named file, pops up a little dialog announcing it's quitting, and then after sleeping for three seconds, the app quits. 

Further interactive testing shows that VS Code responds to the "reopen," "activate," and "show preferences" Apple Events as well. 

It's likely that Electron is picking up this functionality, even though It's not apparently documented, as part of being built on the Cocoa API's - these core events are all baked into the Cocoa framework. In other words, VS Code, and more fundamentally Electron, gets this for free. 

If this functionality is NOT documented, may I suggest a documentation enhancement? And if it is, can someone post a link? 

One additional suggestion. While implementing a FULL AppleScript API is a huge undertaking, would it be too much to implement an AppleScript bridge to an existing Electron scripting API? Tcl/Tk implements a "do script" AppleScript command that lets external apps call into a Tcl/Tk application via the AppleScript bridge, and the script command is then executed by the Tcl interpreter: 

`tell app "Wish"
     do script "puts \"hello world\"
end tell`

The output "hello world" is returned to the calling application. 

Implementing similar functionality in Electron that hooks into one or more of the "executeJavaScript" API's could potentially add a rich IPC mechanism to Electron apps on macOS and improve their ability to interact with other Mac apps.

**Comment by Coder-256** [comment]
The built-in commands you're referencing are [documented in the AppleScript Language Guide](https://developer.apple.com/library/archive/documentation/AppleScript/Conceptual/AppleScriptLangGuide/reference/ASLR_cmds.html); most of them are also part of the Standard Suite you can access using Script Editor from `File > Open Dictionary...`. Most browsers including Chrome and Safari already have an "execute JavaScript" action but now most browsers require it to be enabled from the developer menu, due to obvious security issues (any unsandboxed local app can use AppleScript, and thus could access all your data from that browser). For that reason I think it's best to follow suit and not enable full JavaScript control through AppleScript by default.

Considering it's been nearly 7 years with no progress since I opened this issue, at this point I'm just going to try and implement full, correct scripting dictionary support. Not totally sure if it's feasible but it's worth a shot.

**Comment by andrewschreiber** [comment]
Curious what you found on scripting electron @Coder-256!

**Comment by colinjcahill** [comment]
If @Coder-256 or any other brave soul is interested in building the bridge discussed here I'd love to be the first to donate to the cause.  This is becoming an increasingly vexing programming gap.

**Comment by Coder-256** [comment]
It should be possible, but I just don't have the time to do something this large; it's a much more difficult project than it seems. The main challenge is the fact that you can't really use Cocoa script commands without writing and loading your own application-specific Swift/Objective-C code (bit of a no-no for Electron), so it would be best to handle the raw Apple Events themselves using a JS API. The hardest parts of that will likely be making a good JS wrapper around either the Cocoa/Objective-C APIs (e.g. `NSAppleEventDescriptor`) and/or the C API (e.g. `AEGetParamDesc`, etc.), and figuring out how to actually read the Apple Events that AppleScript generates.

The starting point for getting these events would probably be `-[NSAppleEventManager setEventHandler:andSelector:forEventClass:andEventID:]` ([as suggested here](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ScriptableCocoaApplications/SApps_handle_AEs/SAppsHandleAEs.html#//apple_ref/doc/uid/20001239-1134389)).

Good resources for this are a bit hard to find, partly because Apple Events are a bit old and not very commonly used. Some good resources are:
- The [Cocoa Scripting Guide](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ScriptableCocoaApplications/SApps_intro/SAppsIntro.html): discusses scriptable applications. This is helpful, but some sections don't apply if you aren't using the Cocoa "script command" mechanism which requires custom classes.
- [The Objective-C/Swift API for AppleScript and Apple Events](https://developer.apple.com/documentation/foundation/scripting_support?language=objc). You might still need to use the C API for some things.
- [The C API for Apple Events](https://developer.apple.com/documentation/coreservices/apple_events?language=objc).
- [Introduction to Apple Events Programming Guide](https://developer.apple.com/library/archive/documentation/AppleScript/Conceptual/AppleEvents/intro_aepg/intro_aepg.html): explains a lot about Apple Events and how to use them with the C API.
- [Introduction to Apple Events](https://developer.apple.com/library/archive/documentation/mac/pdf/Interapplication_Communication/Intro_to_Apple_Events.pdf): seemingly an older version of the above, discusses Apple Events in great depth and detail.

Best wishes to whoever takes this on! Let me know if you have any questions and I'll do my best to answer.

**Comment by andrewschreiber** [comment]
Perhaps there is an opportunity here for a "bolt-on" solution in the form of an Electron package. AppleScript is one way to enable two-directional communication between local Electron applications, but not the only way. Maybe it'd be better to have a cross-platform approach here as well.

Lets call the package idea `electron-scriptable`, which would basically be a local [PubSub](https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern) folder.

- Looks for a folder at `~/.electron-scriptable`, creating it if it doesn't exist
- Makes a folder for the Electron app with a path like `~/.electron-scriptable/com.bundle.id/`
- Register to watch for updates in `main.js` (uses fs.watch at a customizable interval)
- Command to create a `schema_v1.json` based on your watchers (probably manual at first)
- For other applications that want to run scripts on your application, they would create a new JSON file in that folder based on the latest schema.
- When a new file is detected in the folder, gets parsed by the function. Perhaps an array of new files since last check.
- Your application can run commands, ignore, etc
- Updates file with response / errors.
- Other application also could have a watcher which detects your update of the JSON file, then run a function

Pros
- Cross-platform
- Could work on older versions of Electron
- Quick to setup
- Could pass arbitrary bi-directionally information as base64
- Other applications could watch as well for composability
- Your application wouldn't have to be open. Could have a message backlog

Cons
- Security. For sensitive data would need to do pinned encryption, which could be reverse engineered
- Spoofable. Other applications might be able to create a fake schema for your application
- Perhaps small battery life / responsiveness trade-off

Overall I think the security concerns are fairly manageable given we're operating on a local machine. That's my sketch, curious what fellow people looking for scriptable Electron apps think!

**Comment by andrewschreiber** [comment]
As an interim solution, using Deep Links / protocols is probably our best bet for at least one-directional communication

https://www.electronjs.org/docs/latest/tutorial/launch-app-from-url-in-another-app

Here's how Raycast extension silently triggers an effect in Things, for example

```
export async function silentlyOpenThingsURL(url: string) {
  const asyncExec = promisify(exec);
  await asyncExec(`open -g "${url}"`);
}

export async function updateTodo(id: string, todoParams: UpdateTodoParams) {
  const { authToken } = getPreferenceValues<Preferences>();

  if (!authToken) throw new Error('unauthorized');

  await silentlyOpenThingsURL(
    `things:///update?${qs.stringify({
      'auth-token': authToken,
      id,
      ...todoParams,
    })}`,
  );
}
```

**Comment by andrewschreiber** [comment]
Had a sketch of an idea on turning URL scheme 1-directional triggered communication into a secure callback experience

Let's say VSCode integrates this package and exposes a scripting interface that lets you, say, trigger code compilation

- vscode://compile_myproject

Now imagine one of the arguments is a freshly generated private key

- vscode://compile_myproject-key_wbdhsk834

Now VSCode, using electron-scriptable, would trigger the command upon receiving the properly formatted scheme. *Then* it would publish the result of that command in the ~/.electron-scriptable folder (as detailed in comment above). The result - compile success - would be encrypted with the private key, so only the original triggering application could parse it.

**Comment by andrewschreiber** [comment]
Following up here with the simplest way to allow your Electron app to be externally scripted via AppleScript. I haven't tested this exactly, but wanted to give the general idea.

In main.ts

``` 

app
  .whenReady()
  .then(async () => {
  
  const p = `applescript-${app.getName().toLowerCase()}`
  
    // Deprecated in Electron 25
    protocol.registerFileProtocol(p, (request, callback) => {
      const textData = request.url.replace(`${p}://', '');
      try {
        // eslint-disable-next-line promise/no-callback-in-promise
        callback(textData);
        console.log("Got applescript call", textData);
      } catch (error) {
        console.error(error);
        // eslint-disable-next-line promise/no-callback-in-promise
      }
    });
  })
  ```
  
  Once this is setup, send data from Applescript as such. When in development, app.getName() will always return `Electron`. Can test in ScriptEditor:
  
  ```
  do shell script "open -g electron://loremipsum"
  ```
  
In the electron main process you will see `loremipsum` printed. 

The maximum string length INT_MAX, so about 2GB worth of data. With base64 encoding/decoding, you should be able to send basically anything.

I still think there's an opportunity for an electron package to standardize and publish the availability around these sort of interactions, but I no longer think a local folder is particularly needed. One could imagine for two-way communication between apps using back and forth URL scheme calling.
